### ISSUE TITLE
Dragging between lists? Nested lists? Wanna build a sortable tree

### ISSUE THREAD

**Issue by lolmaus** [jgwhite/ember-sortable#93]
Here's a prototype I built long ago with jQuery UI Sortable: http://hivemindunit.github.io/hivemind-frontend-prototype/settings/domains/categories2/

I'm now trying to do an identical tree with Ember, and it's a tough nut to crack due to the lack of high-level drag-n-drop abstractions. :(

Is `ember-sortable` able to power a sortable tree like that?

**Comment by jgwhite** [comment]
@lolmaus work is in-progress on the jj-abrams branch :)

**Comment by lolmaus** [comment]
@jgwhite Tell me when it's ready for fiddling with. I'll implement my use case in your dummy app for experimenting with.

**Comment by jgwhite** [comment]
:+1:

**Comment by lolmaus** [comment]
@jgwhite Meanwhile, have you tried out my jQuery UI Sortable tree prototype I linked to above?

Note how all droppable regions are highlighted when you start dragging a node. I'm looking for an Ember addon that would allow building a similar experience. Will be happy to contribute to it once I see it does satisfy my needs.

**Comment by jgwhite** [comment]
@lolmaus yes. I don’t expect to hit that level of functionality on the first try, but I am keeping this use cases in mind to provide extension points.

**Comment by joukevandermaas** [comment]
Is there any progress on this? I'd love to have this functionality.

**Comment by lolmaus** [comment]
I've created a proof-of-concept addon for nested lists: https://github.com/lolmaus/ember-drag-sort

It works, but it requires a lot of time and effort to become usable in apps.

I can invest more time into it, all I need is ~~a swift kick in the butt~ some motivation. We could team up on Slack or something.

**Comment by jgwhite** [comment]
@lolmaus this looks very promising. Sorry progress has been so slow on my end.

@lolmaus @opsb would you folks be interested in a hangout at some point soon to discuss joining forces?

**Comment by lolmaus** [comment]
@jgwhite Poke me on Slack.

**Comment by edeis53** [comment]
@jqwhite just to clarify, does master currently support nesting/children? 

Example model:
```
'items' => [
    ['content' => 'Item # 1', 'id' => 1],
    ['content' => 'Item # 2', 'id' => 2],
    ['content' => 'Item # 3', 'id' => 3],
    ['content' => 'Item # 4 with children', 'id' => 4, 'children' => [
        ['content' => 'Item # 4.1', 'id' => 5],
        ['content' => 'Item # 4.2', 'id' => 6],
        ['content' => 'Item # 4.3', 'id' => 7],
    ]],
],
```

**Comment by jgwhite** [comment]
@edeis53 no, it doesn't.

I explored this space pretty intensely with @scottmessinger. You can see the results in the jj-abrams branch. In the end, the problem space was so vast I completely burnt out on it.

The hard part is designing a set of primitives that support sorting between trees of lists, that doesn't dictate a particular style of nested sorting UI. All whilst supporting animation as we currently do without becoming too complex or expensive.

TL;DR the problem beat me, but I think there's a good solution out there waiting to be discovered.

**Comment by lolmaus** [comment]
What's the status of the `jj-abrams` branch?

**Comment by jgwhite** [comment]
@lolmaus I'm afraid I burnt out working on it. Consider it abandoned.

**Comment by scottmessinger** [comment]
@lolmaus I'm using it right now and it's working great. There are definitely some areas that should be reworked, but I haven't had time to make the changes. Specifically, I think the ability of `SortableNode` (new concept in the `jj-abrams` branch) to be both a `sortable-group` and a `sortable-item` causes quite a bit of confusion. I've found the code backing SortableNode to be really first rate and I'm grateful to @jgwhite for publishing the branch. It's a huge part of our app and it's worked exceptionally well. 

I'd encourage future collaborators to look at using the `SortableNode` code to back the `sortable-group` and `sortable-item` interface from v1 while adding the concept of "connected groups". I'd also encourage keeping `SortableNode` as a public API primitive. For instance, In a tree structure where the sortable-groups and the sortable-items are the same model (like @edeis53 posted) , SortableNode is probably the right API. In our app where groups are pretty distinct from the items inside, `sortable-group` and `sortable-item` are a more helpful interface. 

As for stability, I've definitely found it stable enough to use. Hit me up on the ember slack room if you have any questions.

**Comment by Syzygy24** [comment]
@jgwhite Is something like this supported?

```
{{#sortable-group}}
    {{parent items}}
    {{#if item.children}}
        {{#sortable-group}}
            {{child items}}
        {{/sortable-group}}
    {{/if}}
{{/sortable-group}}
```

I'm trying to get it working where the child items are an entirely independent list other than their tie to the parent, but no luck so far.

**Comment by jgwhite** [comment]
We’ve decided tree-sorting is beyond the scope of ember-sortable as it stands today. Building an API that supports arbitrary modes of tree sorting is a massive undertaking and for now ember-sortable best serves the community by performing its simple task dependably.
